app/assets/javascripts/darkswarm/controllers/hub_node_controller.js.coffee

```diff
@@ -1,9 +1,44 @@
-Darkswarm.controller "HubNodeCtrl", ($scope, HashNavigation, Navigation, $location, $templateCache, CurrentHub) ->
-  $scope.toggle = (e) ->
-    HashNavigation.toggle $scope.hub.hash if !angular.element(e.target).inheritedData('is-link')
+Darkswarm.controller "HubNodeCtrl", ($scope, HashNavigation, CurrentHub, $http, $timeout) ->
+  $scope.shopfront_loading = false
+  $scope.enterprise_details = []
 
+  $timeout ->
+    if $scope.open()
+      $scope.load_shopfront()
+
+  # Toggles shopfront tabs open/closed. Fetches enterprise details from the api, diplays them and adds them
+  # to $scope.enterprise_details, or simply displays the details again if previously fetched
+  $scope.toggle = (event) ->
+    if $scope.open()
+      $scope.toggle_tab(event)
+      return
+
+    if $scope.enterprise_details[$scope.hub.id]
+      $scope.hub = $scope.enterprise_details[$scope.hub.id]
+      $scope.toggle_tab(event)
+      return
+
+    $scope.load_shopfront(event)
+
+  $scope.load_shopfront = (event=null) ->
+    $scope.shopfront_loading = true
+    $scope.toggle_tab(event)
+
+    $http.get("/api/enterprises/" + $scope.hub.id + "/shopfront")
+      .success (data) ->
+        $scope.shopfront_loading = false
+        $scope.hub = data
+        $scope.enterprise_details[$scope.hub.id] = $scope.hub
+      .error (data) ->
+        console.error(data)
```

Comment:
(1) data loading should not be part of the controller.
it should be extracted to a separate file, like a resource, for example.

(2) also, I think we should continue the separation of darkswarm in domains like admin (I remember Rob mentioning this somewhere).
so, instead of 
app/assets/javascripts/darkswarm/controllers/hub_node_controller.js.co
we would have 
app/assets/javascripts/darkswarm/shops/hub_node_controller.js.coffee

and then this data loading code could be extracted to a new file:
app/assets/javascripts/darkswarm/shops/enterprises_resource.js.coffee

if you only implement suggestion (1) above, this code could be extracted to app/assets/javascripts/darkswarm/services/enterprise_resource.js.coffee (not an api endpoint)
or we could create a new file for api/enterprise
app/assets/javascripts/darkswarm/resources/api/enterprise_resource.js.coffee (for api endpoints)

this is important as we would make a lot more explicit the repeated code in darkswarm and admin:
app/assets/javascripts/admin/resources/resources/enterprise_resource.js.coffee

I think we need to agree on a location for these things. Otherwise, the existing mess will just keep growing.
What do you think?

Reply:
Ah yes, I was planning on extracting these to a new resource when it was all working correctly. :+1:

Reply:
better to address in a separate PR IMO. I'll be much cleaner.